jupyter-repl.el

```diff
@@ -962,6 +962,66 @@ buffer to display TEXT."
           (font-lock-prepend-text-property
            beg end 'font-lock-face 'jupyter-repl-traceback)))))))
 
+(defun jupyter-repl-history--match-input (regexp arg)
+  "Return the index of the ARGth REGEXP match.
+Or nil, on failure. If ARG is positive, search backward from the most
+recent history element. If negative, search forward through items
+previously visited during this input session. If ARG is zero, do
+nothing."
+  ;; Adapted from `comint-previous-matching-input-string-position'
+  (let* ((r jupyter-repl-history)
+         (direction (if (> arg 0) +1 -1))
+         (i (if (= direction -1) 0 -1)) ; adjust for initial increment
+         (failed (zerop arg))
+         ;;
+         code)
+    ;; Search ARG times
+    (while (not (or failed (zerop arg)))
+      (while (not (or (setq code (ring-ref r (cl-incf i direction))
+                            failed (eq 'jupyter-repl-history code))
+                      (string-match-p regexp code))))
+      (setq arg (- arg direction)))
+    (unless failed i)))
+
+(defun jupyter-repl-history-previous-matching (&optional n)
+  "Search input history for the input pending before point.
+On success, replace the current input with the matching code element
+while preserving and returning point. Ding on failure. If N is negative,
+find the Nth next match; if positive, the Nth previous. If N is zero or
+nil, pretend it's one."
+  ;; Adapted from: `comint-previous-matching-input-from-input' and friends
+  (interactive "p")
+  (when (or (null n) (zerop n)) (setq n 1))
+  (let ((opoint (point))
+        (r jupyter-repl-history)
```

Comment:
Can we get rid of this `r` variable,  we can just use `jupyter-repl-history`.

Reply:
Removed the variable in both functions.